### ISSUE TITLE
Recursive annotations triggers a StackOverflowError

### ISSUE THREAD

**Issue by marvelous** [eclipse-ee4j/eclipselink#63]
When an annotation class annotated with itself gets scanned by EclipseLink (`<exclude-unlisted-classes>false</exclude-unlisted-classes>`), it enters into a recursive loop which triggers a `StackOverflowError` where `org.eclipse.persistence.internal.jpa.metadata.accessors.objects.MetadataAnnotatedElement.getAnnotation(MetadataAnnotatedElement.java:215)` calls itself.

```
@Self
public @interface Self {}
```

It happens in payara-micro 5 (EclipseLink 2.7.1.qualifier), but it didn't happen in payara-micro 4. A sample WAR can be found in payara/Payara#2608.

**Comment by marvelous** [comment]
Here is a Java SE sample: [sscce.zip](https://github.com/eclipse-ee4j/eclipselink/files/1893823/sscce.zip)

**Comment by rieonke** [comment]
so can we just ignore the annotations which self annotated?
I ignored them at https://github.com/eclipse-ee4j/eclipselink/blob/a985f91916b85540209553514cafd9f88718c8c6/jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/metadata/accessors/objects/MetadataAsmFactory.java#L373-L385

```java
        @Override
        public void visitEnd() {
            if (this.element != null && this.annotation != null && !this.annotation.getName().equals(this.element.getName())) {
                if (this.annotation.isMeta()) {
                    this.element.addMetaAnnotation(this.annotation);
                } else {
                    this.element.addAnnotation(this.annotation);
                }
            }
        }
```
then it works fine.

@lukasj

**Comment by marvelous** [comment]
I doubt it will be enough to catch mutual recursion like

```
@Self2
public @interface Self1 {}
```
```
@Self1
public @interface Self2 {}
```

**Comment by cristhiank** [comment]
I created a bug in Bugzilla also because it's not clear if the project uses Github as the bug tracker: https://bugs.eclipse.org/bugs/show_bug.cgi?id=535250

**Comment by Tomas-Kraus** [comment]
Well, that's a question. There are 2 options how to handle it:
1. break the cycle while reading classes (fix proposed here)
2. break the cycle in the MetadataAnnotation getAnnotation(String annotation) method which is causing stack overflow

(1.) means that the last class of the cycle will be missing in the metamodel and I think that it's not good even when we know that something is going wrong here
I would prefer to rely on a.getName() being unique in the whole stack chain of getAnnotation calls and break it on 1st repeated a.getName().

**Comment by cristhiank** [comment]
Hi @Tomas-Kraus , thanks for your revision. 
I'm new to the eclipselink codebase and I'm starting to understand it's structure. 
I'm doing some tests to try to solve it as you say in the 2nd option.

**Comment by cristhiank** [comment]
@Tomas-Kraus I added an alternative solution with your recommendations. Looking forward to your revision.

**Comment by Tomas-Kraus** [comment]
OK. I rewrote your 1st option to make it more simple: https://github.com/eclipse-ee4j/eclipselink/pull/133
Also added 2 jUnits to verify A->A and A->B->C->A scenarios.
@lukasj I would also like to hear your opinion.

**Comment by Tomas-Kraus** [comment]
Last commit "Bug 535250 - Fix in getAnnotation processing" contains PR 109 change rewritten to be able to detect any cycle. Unfortunately this added some additional complexity there.
Thisi is why I may change my mind and go with original proposal to fix it in initial metadata processing - MetadataAnnotationVisitor#visitEnd().

**Comment by marvelous** [comment]
Seems fixed in master and 2.7 branch. Thanks!
